zanata-war/src/main/java/org/zanata/liquibase/custom/ChangePositionalResIdToContentHash.java

```diff
@@ -0,0 +1,224 @@
+    private PreparedStatement selectDocumentsOfType;
+    private PreparedStatement selectTextFlows;
+
+    private int totalDocs = 0;
+    private int totalTextFlows = 0;
+
+    @Override
+    public void execute(Database database) throws CustomChangeException {
+
+        try {
+            prepareStatements((JdbcConnection) database.getConnection());
+            // could use try-with-resources to define the PreparedStatements here
+            // but making them fields allows for cleaner code.
+            try {
+                for (DocumentType type : documentTypes) {
+                    migrateResId(type);
+                }
+            } finally {
+                closeStatements();
+            }
+        } catch (DatabaseException | SQLException e) {
+            throw new CustomChangeException(e);
+        }
+        // connection is not closed in a finally clause since doing so causes liquibase to throw exceptions.
+    }
+
+    private void prepareStatements(JdbcConnection conn) throws DatabaseException {
+        countDocumentsOfType = conn.prepareStatement(countDocumentsOfTypeSql,
+                ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY);
+        selectDocumentsOfType = conn.prepareStatement(selectDocumentsOfTypeSql,
+                ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY);
+        selectTextFlows = conn.prepareStatement(selectTextFlowsSql,
+                ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_UPDATABLE);
+    }
+
+    private void closeStatements() throws SQLException {
+        countDocumentsOfType.close();
+        selectDocumentsOfType.close();
+        selectTextFlows.close();
+    }
+
+    /**
+     * Update resId for all text flows in documents of the given type.
+     *
+     * @param type type to migrate
+     * @throws SQLException
+     */
+    private void migrateResId(DocumentType type) throws SQLException {
+        long docsOfType;
+        long docsProcessed = 0;
+        countDocumentsOfType.setString(1, type.name());
+        try (ResultSet countOfType = countDocumentsOfType.executeQuery()) {
+            if (countOfType.next()) {
+                docsOfType = countOfType.getLong(1);
+                log.info("processing " + docsOfType + " documents of type " + type.name());
+            } else {
+                // unexpected error, but does not have to destroy everything.
+                throw new SQLException("Count query returned no rows.");
+            }
+        }
+
+        selectDocumentsOfType.setString(1, type.name());
+        try (ResultSet documentIds = selectDocumentsOfType.executeQuery()) {
+            while (documentIds.next()) {
+                Long documentId = documentIds.getLong(1);
+                migrateResIdForDocument(documentId);
+                docsProcessed++;
+                if (docsProcessed % 100 == 0) {
+                    log.info("    processed " + docsProcessed + " of " + docsOfType);
```

Comment:
@seanf I wanted to avoid it logging progress for types that have few documents since that would just be a lot of noise with no value. I guess I could add after the `while` loop:

```
if (docsProcessed > 100) {
    log.info(" processed " + docsProcessed + " of " + docsOfType);
}
```

(or just log that unconditionally).